delegate-app/Main.hs

```diff
@@ -152,12 +165,55 @@ runDelegateServer conf = do
     pure q
   delegateBroadCastChan <- liftIO . atomically $ newTChan
 
-  _ <-
-    liftIO . forkIO $
-      runServer (show $ dlgt'host conf) (fromIntegral $ dlgt'port conf) $
-        runWithTracer' tracer . delegateServerApp (dlgt'ping conf) delegateInputChan delegateBroadCastChan
-
-  execDelegateRunnerT $ mkRunner (dlgt'tick conf) delegateInputChan delegateBroadCastChan
+  let startWorker :: ThreadSort -> IO () -> IO ()
+      startWorker thread act = do
+        traceWith tracer $ StartThread thread
+        act
+        traceWith tracer $ CancelThread thread
+
+      wsServer, delegateRunner, queueAuctionPhases :: IO ()
+      wsServer =
+        runServer (show $ dlgt'host conf) (fromIntegral $ dlgt'port conf) $
+          runWithTracer' tracer . delegateServerApp (dlgt'ping conf) delegateInputChan delegateBroadCastChan
+
+      delegateRunner =
+        runWithTracer' tracer $ execDelegateRunnerT $ mkRunner (dlgt'tick conf) delegateInputChan delegateBroadCastChan
+
+      queueAuctionPhases = runWithTracer' tracer $ mbQueueAuctionPhases delegateInputChan delegateBroadCastChan
+
+  liftIO $
+    mapConcurrently_
+      (uncurry startWorker)
+      [ (WebsocketThread, wsServer)
+      , (QueueAuctionStageThread, queueAuctionPhases)
+      , (DelegateRunnerThread, delegateRunner)
+      ]
+
+mbQueueAuctionPhases :: TQueue DelegateInput -> TChan DelegateResponse -> DelegateTracerT IO ()
+mbQueueAuctionPhases reqq =
+  liftIO . atomically . (dupTChan >=> getTerms)
+    >=> liftIO . (traceQueueEvent *> queueCurrentStage)
+  where
+    traceQueueEvent :: AuctionTerms -> DelegateTracerT IO ()
+    traceQueueEvent = trace . QueueAuctionPhaseEvent . ReceivedAuctionSet
+
+    getTerms :: TChan DelegateResponse -> STM AuctionTerms
+    getTerms chan =
+      readTChan chan >>= \case
+        AuctionSet terms -> pure terms
+        _ -> getTerms chan
+
+    queueCurrentStage :: AuctionTerms -> IO ()
```

Comment:
Name is misleading, this is forever-running thread.

Reply:
This should be somehow marked by name. `queueCurrentStage` sounds like one-action function.